armory/utils/metrics.py

```diff
@@ -431,6 +431,192 @@ def _check_object_detection_input(y, y_pred):
+    assert box_1[3] >= box_1[1]
+    assert box_2[3] >= box_2[1]
+
+    if sum([mean < 1 and mean >= 0 for mean in [box_1.mean(), box_2.mean()]]) == 1:
+        logger.warning(
+            "One set of boxes appears to be normalized while the other is not"
+        )
+
+    # Determine coordinates of intersection box
+    x_left = max(box_1[1], box_2[1])
+    x_right = min(box_1[3], box_2[3])
+    y_top = max(box_1[0], box_2[0])
+    y_bottom = min(box_1[2], box_2[2])
+
+    intersect_area = max(0, x_right - x_left) * max(0, y_bottom - y_top)
+    if intersect_area == 0:
+        return 0
+
+    box_1_area = (box_1[3] - box_1[1]) * (box_1[2] - box_1[0])
+    box_2_area = (box_2[3] - box_2[1]) * (box_2[2] - box_2[0])
+
+    iou = intersect_area / (box_1_area + box_2_area - intersect_area)
+    assert iou >= 0
+    assert iou <= 1
+    return iou
+
+
+def object_detection_mAP(list_of_ys, list_of_y_preds):
+    """
+    Mean average precision for object detection. This function returns a dictionary
+    mapping each class to the average precision (AP) for the class. The mAP can be computed
+    by taking the mean of the AP's across all classes.
+
+    This metric is computed over all evaluation samples, rather than on a per-sample basis.
+    """
+
+    IOU_THRESHOLD = 0.5
+    # Precision will be computed at recall points of 0, 0.1, 0.2, ..., 1
+    RECALL_POINTS = np.linspace(0, 1, 11)
+
+    # Converting all boxes to a list of dicts (a list for predicted boxes, and a
+    # separate list for ground truth boxes), where each dict corresponds to a box and
+    # has the following keys "img_idx", "label", "box", as well as "score" for predicted boxes
+    pred_boxes_list = []
+    gt_boxes_list = []
+    for img_idx, (y, y_pred) in enumerate(zip(list_of_ys, list_of_y_preds)):
+        for gt_box_idx in range(len(y["labels"][0].flatten())):
+            label = y["labels"][0][gt_box_idx]
+            box = y["boxes"][0][gt_box_idx]
+
+            gt_box_dict = {"img_idx": img_idx, "label": label, "box": box}
+            gt_boxes_list.append(gt_box_dict)
+
+        for pred_box_idx in range(len(y_pred["labels"].flatten())):
+            label = y_pred["labels"][pred_box_idx]
+            box = y_pred["boxes"][pred_box_idx]
+            score = y_pred["scores"][pred_box_idx]
+
+            pred_box_dict = {
+                "img_idx": img_idx,
+                "label": label,
+                "box": box,
+                "score": score,
+            }
+            pred_boxes_list.append(pred_box_dict)
+
+    # Union of (1) the set of all true classes and (2) the set of all predicted classes
+    set_of_class_ids = set([i["label"] for i in gt_boxes_list]) | set(
+        [i["label"] for i in pred_boxes_list]
+    )
+
+    # Remove the class ID that corresponds to a physical adversarial patch in APRICOT
+    # dataset, if present
+    # TODO: retrieve this ID from datasets.py rather than hardcoding it
+    set_of_class_ids.discard(-10)
```

Comment:
Saw this was left as a TODO.  May want to replace this before merging to prevent confusion down the line.

Reply:
done